### PR TITLE
Relayer commits

### PR DESCRIPTION
**Pull request by misko9**

### CODE REVIEW

File: rust/chains/hyperlane-cosmos-modules/src/interchain_gas.rs [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+
+use crate::{Signer, ConnectionConf, CosmosProvider};
+
+#[derive(Debug)]
+pub struct CosmosInterchainGasPaymaster {
+    conf: ConnectionConf,
+    domain: HyperlaneDomain,
+    address: H256,
+    signer: Signer,
+    provider: Box<CosmosProvider>,
```
Comment (KyleMoser) [inline rust/chains/hyperlane-cosmos-modules/src/interchain_gas.rs]: Curious - how come this is Boxed?

File: rust/chains/hyperlane-cosmos-modules/src/interchain_gas.rs [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+            gas_amount: U256::default(),
+        };
+        for attr in attrs {
+            let key = attr.key.as_str();
+            let value = attr.value.as_str();
+
+            info!("Attr: {:?}", key);
+            match key {
+                "messageid" => {
+                    res.message_id = H256::from_slice(hex::decode(value.trim_start_matches("0x")).unwrap().as_slice())
```
Comment (KyleMoser) [inline rust/chains/hyperlane-cosmos-modules/src/interchain_gas.rs:91]: Is the `messageid` guaranteed to start with 0x? I looked up `trim_start_matches` and it will not behave safely if the `messageid` doesn't start with 0x (returns None in that case). Some of the IGP test code uses messageids without the 0x prefix, I thought that was the expected input for some reason. If not, we can change it in the IGP test code.

File: rust/chains/hyperlane-cosmos-modules/src/interchain_security_module.rs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+                grpc_url: "http://127.0.0.1:45897".to_string(),
+                rpc_url: "".to_string(),
+                chain_id: "".to_string(),
+            },
+            ContractLocator { 
+                domain: &HyperlaneDomain::Unknown {
+                    domain_id: 0,
+                    domain_name: "CosmosTest".to_string(),
+                    domain_type: HyperlaneDomainType::LocalTestChain,
+                    domain_protocol: HyperlaneDomainProtocol::Ethereum,
```
Comment (KyleMoser) [inline rust/chains/hyperlane-cosmos-modules/src/interchain_security_module.rs]: What does the protocol being ethereum signify and why would ours use this protocol?

**Review by KyleMoser — COMMENTED** [review]
Biggest comment is that functions need comments. I know the protocol itself has extensive documentation, and that this code is just implementing interfaces required by the protocol - but it's hard to follow/understand what each piece does without function level comments.

File: rust/chains/hyperlane-cosmos-modules/src/interchain_gas.rs (hunk 1 above)
Comment (misko9) [reply rust/chains/hyperlane-cosmos-modules/src/interchain_gas.rs]: Boxing the provider is probably the lowest hanging fruit in relieving stack memory pressure and is the pattern for the HyperlaneChain provider trait.

File: rust/chains/hyperlane-cosmos-modules/src/interchain_security_module.rs (hunk 3 above)
Comment (misko9) [reply rust/chains/hyperlane-cosmos-modules/src/interchain_security_module.rs]: This was a remnant from early unit testing and didn't matter for that test case, only needed to construct the new object. A majority of the unit testing required a specific environment set up to test that piece before e2e was functional. Now that e2e exists, I have removed those unit/integration tests that require manual env/test case setup.

File: rust/chains/hyperlane-cosmos-modules/src/interchain_gas.rs (hunk 2 above)
Comment (misko9) [reply rust/chains/hyperlane-cosmos-modules/src/interchain_gas.rs:91]: Good catch. Given that the IGP event attribute is output based on the MsgPayForGas input, it is possible that it doesn't have the 0x. I'd propose we change MsgPayForGas's MessageId to by bytes instead of string. Then, the cli tx can support both if we want and we'd have consistency in how the payment is stored in the KVStore (already a byte slice) along with outputting the event attribute consistently. I'd still like it to output with the 0x prefix to be consistent with the mailbox and ISM modules.

**Comment by misko9** [comment]
Added some more functional level comments.
